waiter/src/waiter/auth/spnego.clj

```diff
@@ -83,32 +86,55 @@
    will be run, otherwise the handler will not be run and 401
    returned instead.  This middleware doesn't handle cookies for
    authentication, but that should be stacked before this handler."
-  [request-handler password]
-  (fn require-gss-handler [{:keys [headers] :as req}]
+  [request-handler ^ExecutorService thread-pool password]
+  (fn require-gss-handler [{:keys [headers] :as request}]
     (let [waiter-cookie (auth/get-auth-cookie-value (get headers "cookie"))
           [auth-principal _ :as decoded-auth-cookie] (auth/decode-auth-cookie waiter-cookie password)]
       (cond
         ;; Use the cookie, if not expired
         (auth/decoded-auth-valid? decoded-auth-cookie)
         (let [auth-params-map (auth/auth-params-map auth-principal)
               request-handler' (middleware/wrap-merge request-handler auth-params-map)]
-          (request-handler' req))
+          (request-handler' request))
         ;; Try and authenticate using kerberos and add cookie in response when valid
-        (get-in req [:headers "authorization"])
-        (let [^GSSContext gss_context (gss-context-init)
-              token (do-gss-auth-check gss_context req)]
-          (if (.isEstablished gss_context)
-            (let [principal (gss-get-princ gss_context)
-                  user (first (str/split principal #"@" 2))
-                  resp (auth/handle-request-auth request-handler req user principal password)]
-              (log/debug "added cookies to response")
-              (if token
-                (if (map? resp)
-                  (rr/header resp "WWW-Authenticate" token)
-                  (async/go
-                    (rr/header (async/<! resp) "WWW-Authenticate" token)))
-                resp))
-            (response-401-negotiate)))
+        (get-in request [:headers "authorization"])
+        (let [current-correlation-id (cid/get-correlation-id)
+              response-chan (async/promise-chan)
+              timer-context (timers/start (metrics/waiter-timer "core" "spnego" "throttle-delay"))]
+          ;; launch task that will populate the response in response-chan
```

Comment:
Could also metric the current size of the thread pool here. (Or rather, metric the number of free threads; getMaxPoolSize()-getActiveCount().... If it gets to zero, then we know we're bottlenecking here. Also metric'ing the blockingQueueSize is possible, but IMO, less important. I'd rather know we're backlogging and how often, not how long the backlog is.

Reply:
Added metrics for active and max threads. Also added metrics for the number of pending tasks in the queue.

Reply:
I'd metric not the number of active threads, but the available threads. (max-active). 
The absolute number of active threads isn't an interesting number, its when the number gets close the the max, specifically when max-active gets low, that we care about.

The cognitive load of 'when this number gets close to zero there's a problem', is simpler than 'when the number gets close to another number, there's a problem'.  It'll also graph nicer, and we can put an alert in when, say, we get to available threads < 2 for more than X times per minute.

Reply:
Done.